### ISSUE TITLE
FQDN / shortname mismatch

### ISSUE THREAD

**Issue by nickperry** [puppetlabs/puppetlabs-kubernetes#60]
I am running 1.0.2 of the module. My cluster looks like this:

```
kubectl get nodes
NAME                                  STATUS    ROLES     AGE       VERSION
foo001.mydomain.com   Ready     <none>    4h        v1.9.2
foo002.mydomain.com   Ready     <none>    4h        v1.9.2
foo003.mydomain.com   Ready     <none>    4h        v1.9.2
```

These two blocks of code in kube_addons.pp are causing my Puppet runs to fail because they expect my nodes to registered by their short name, rather than FQDN.

```
 77   if $controller {
 78     exec { 'Assign master role to controller':
 79       command => "kubectl label node ${::hostname} node-role.kubernetes.io/master=",
 80       unless  => "kubectl describe nodes ${::hostname} | tr -s ' ' | grep 'Roles: master'",
 81     }
```

```
 94       exec { 'Taint master node':
 95         command => "kubectl taint nodes ${::hostname} key=value:NoSchedule",
 96         onlyif  => 'kubectl get nodes',
 97         unless  => "kubectl describe nodes ${::hostname} | tr -s ' ' | grep 'Taints: key=value:NoSchedule'"
 98       }
```

```
Error: kubectl label node foo001 node-role.kubernetes.io/master= returned 1 instead of one of [0]
Error: /Stage[main]/Kubernetes::Kube_addons/Exec[Assign master role to controller]/returns: change from notrun to 0 failed: kubectl label node foo001 node-role.kubernetes.io/master= returned 1 instead of one of [0]
```

```
Notice: /Stage[main]/Kubernetes::Kube_addons/Exec[Taint master node]/returns: Error from server (NotFound): nodes "foo001" not found
Error: kubectl taint nodes foo001 key=value:NoSchedule returned 1 instead of one of [0]
Error: /Stage[main]/Kubernetes::Kube_addons/Exec[Taint master node]/returns: change from notrun to 0 failed: kubectl taint nodes foo001 key=value:NoSchedule returned 1 instead of one of [0]
```

**Comment by scotty-c** [comment]
@nickperry let me raise a ticket internally as a bug for this

**Comment by scotty-c** [comment]
@nickperry This ticket will make it into our next sprint, just an FYI

**Comment by davejrt** [comment]
@nickperry are you changing anything else in the params to register your nodes by their fqdn?

**Comment by nickperry** [comment]
@davejrt I used wildcard certs and used kubernetes::bootstrap_controller_ip: "<my_master_hostname_spec>.%{::domain}" (FQDN) instead of an IP address.

This was necessary to work around to work around https://github.com/puppetlabs/puppetlabs-kubernetes/issues/52 .

**Comment by davejrt** [comment]
@nickperry can you please pass me the output of `kubectl get nodes -o yaml`...I'm mostly interested in the top portion with the labels and the node name.

**Comment by nickperry** [comment]
@davejrt how come the fix got reverted? Worked well for me.

Here is the output you asked for:

apiVersion: v1
items:
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      node.alpha.kubernetes.io/ttl: "0"
      volumes.kubernetes.io/controller-managed-attach-detach: "true"
    creationTimestamp: 2018-03-08T16:46:40Z
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/os: linux
      kubernetes.io/hostname: fooxkub01mst001.sub.domain.com
    name: fooxkub01mst001.sub.domain.com
    namespace: ""
    resourceVersion: "33907"
    selfLink: /api/v1/nodes/fooxkub01mst001.sub.domain.com
    uid: 4671d09c-22f0-11e8-8b01-0050568271ce
  spec:
    externalID: fooxkub01mst001.sub.domain.com
  status:
    addresses:
    - address: 10.101.3.21
      type: InternalIP
    - address: fooxkub01mst001.sub.domain.com
      type: Hostname

**Comment by davejrt** [comment]
@nickperry it was reverted because it caused an issue if the node name wasnt your fqdn which is the default behaviour of the module. 

Im curious to know how your node name is the fqdn if the hostname and fqdn facts are different

**Comment by nickperry** [comment]
I'm not sure why mine is behaving differently. How would you feel about the role / taint logic supporting both hostname and fqdn, by determining which to use from the output of a kubectl get nodes?

**Comment by davejrt** [comment]
@nickperry we definitely are not opposed to that, in fact @scotty-c have been spitballing this idea internally. We are more just interested in why your setup specifically behaves differently. I suppose we are just trying to weed out the edge cases a bit more.

But definitely your suggestion of supporting both is something we're open to. It's just a case of a determining why yours picks up the fqdn vs the hostname fact. As we still need to evaluate those.

**Comment by nickperry** [comment]
Happy to provide any additional information that might be useful to figure out why. Really, the only place where I've gone off piste at all is (as mentioned above) my using an fqdn for kubernetes::bootstrap_controller_ip. But I really need to be able to do that.

**Comment by davejrt** [comment]
Thanks Nick, I'm going to play around a little more with the kream environment and see if I can some up with something more concrete, but I'm more than happy to implement the hostname and fqdn in final solution once I track down where the discrepancy is. 

Appreciate your patience and feedback.

**Comment by davejrt** [comment]
fixed in PR #76
